src/raftstore/store/worker/read.rs

```diff
@@ -0,0 +1,895 @@
+            Task::Read(ref msg) => write!(f, "localreader Task::Msg {:?}", msg),
+            Task::Update(ref progress) => write!(f, "localreader Task::Update {:?}", progress),
+            Task::Destroy(region_id) => write!(f, "localreader Task::Destroy region {}", region_id),
+        }
+    }
+}
+
+fn handle_busy(cmd: StoreMsg) {
+    let mut err = errorpb::Error::new();
+    err.set_message(RAFTSTORE_IS_BUSY.to_owned());
+    let mut server_is_busy = errorpb::ServerIsBusy::new();
+    server_is_busy.set_reason(RAFTSTORE_IS_BUSY.to_owned());
+    err.set_server_is_busy(server_is_busy);
+    let mut resp = RaftCmdResponse::new();
+    resp.mut_header().set_error(err);
+
+    let read_resp = ReadResponse {
+        response: resp,
+        snapshot: None,
+    };
+
+    match cmd {
+        StoreMsg::RaftCmd { callback, .. } => callback.invoke_read(read_resp),
+        StoreMsg::BatchRaftSnapCmds {
+            batch, on_finished, ..
+        } => on_finished.invoke_batch_read(vec![Some(read_resp.clone()); batch.len()]),
+        other => panic!("unexpected cmd {:?}", other),
+    }
+}
+
+pub struct LocalReader<C: Sender<StoreMsg>> {
+    store_id: u64,
+    kv_engine: Arc<DB>,
+    metrics: ReadMetrics,
+    // region id -> ReadDelegate
+    delegates: HashMap<u64, ReadDelegate>,
+    // A channel to raftstore.
+    ch: C,
+    tag: String,
+}
+
+impl LocalReader<mio::Sender<StoreMsg>> {
+    pub fn new<T, P>(store: &Store<T, P>) -> Self {
+        let mut delegates =
+            HashMap::with_capacity_and_hasher(store.get_peers().len(), Default::default());
+        for (&region_id, p) in store.get_peers() {
+            let delegate = ReadDelegate::from_peer(p);
+            debug!(
+                "{} init ReadDelegate for peer {:?}",
+                delegate.tag, delegate.peer
+            );
+            delegates.insert(region_id, delegate);
+        }
+        let store_id = store.store_id();
+        LocalReader {
+            delegates,
+            store_id,
+            kv_engine: store.kv_engine(),
+            ch: store.get_sendch().into_inner(),
+            metrics: ReadMetrics::default(),
+            tag: format!("[store {}]", store_id),
+        }
+    }
+}
+
+impl<C: Sender<StoreMsg>> LocalReader<C> {
+    fn redirect(&mut self, cmd: StoreMsg) {
+        debug!("{} localreader redirect {:?}", self.tag, cmd);
+        self.metrics.raft_cmd_rejected += 1;
+        match self.ch.send(cmd) {
+            Ok(()) => (),
+            Err(NotifyError::Full(cmd)) => {
+                self.metrics.rejected_by_channel_full += 1;
+                handle_busy(cmd)
+            }
+            Err(err) => {
+                panic!("localreader redirect failed: {:?}", err);
+            }
+        }
+    }
+
+    fn pre_propose_raft_command(&mut self, req: &RaftCmdRequest) -> Result<bool> {
+        // Check store id.
+        if let Err(e) = util::check_store_id(req, self.store_id) {
```

Comment:
The process seems duplicated with store, can it be reused?

Reply:
I tried, but there are different logic and error handling, eg, `per_propose_raft_command` also handles status requests and epoch error.